### ISSUE TITLE
object-level / flattened caching

### ISSUE THREAD

**Issue by micimize** [zino-app/graphql-flutter#109]
**Is your feature request related to a problem? Please describe.**
We're currently doing [operation level caching](https://github.com/zino-app/graphql-flutter/blob/next/lib/src/core/query_manager.dart#L124) instead of the [object-level caching](https://www.apollographql.com/docs/react/advanced/caching.html) available in apollo. This means disparate operations on the same resource result in inconsistent cache data.

**Describe the solution you'd like**
The first steps here look to me to be:
* add an optional `dataIdFromObject` equivalent
* traverse results and cache them based on the `dataId`
* replace values in operation cache with references

This would allow mutation results to automatically invalidate/update query results, which would also be a good step towards user friendly optimistic updates.

**Additional context**
While eventually it'd be nice to [inject `__typename` automatically](https://github.com/apollographql/apollo-client/blob/c3561e73957799bbf0d7f87c4a3ebcfee6f40e3a/packages/apollo-utilities/src/transform.ts#L168) the way apollo does, maybe using angel's [graphql_parser](https://github.com/angel-dart/graphql/tree/master/graphql_parser), I think we can punt that to userland initially.

**Comment by HofmannZ** [comment]
@micimize Yeah we're thinking about adding that, just need to find some time to actually do it ;)

**Comment by micimize** [comment]
I can start on this today, if you all haven't already. My implementation plan, again:
* `dataIdFromObject` will be called on each object in the tree.
* If not null, the leaf will be replaced with `['@cache/reference', dataId]`
* references will be which will be saturated whenever the store is queried

**Comment by HofmannZ** [comment]
@micimize Would be great to see another PR coming in. We can definitely should use the `__typename` key. And since we already rely on [graphql_parser](url
https://github.com/angel-dart/graphql/tree/master/graphql_parser) for operation names, we would not be adding any more dependencies.

Adding the `__typename` key would also mean we can align optimistic UI updates with the way Apollo implements them.

Looking forward to collaborate on this feature.

**Comment by twpochuan** [comment]
Good to see you guys are working on this.

I tried to implement the normalized cache but I got some problems to make it generalized.
I don't know the problems I ran into will be helpful or not. 
Here are the two main problems I faced:

The first problem is that not all of the given objects have id field, which may causes duplicated cache for the same object.

The second problem is about pagination.
For example, for relay type of pagination, there is a Connection type.
https://facebook.github.io/relay/graphql/connections.htm
If a user want to get a query with {first: 20} or {first: 5} from the cache, however, the user only queried from network with arguments: {first: 10} and {first: 10, after: 'a cursor}. Without taking pagination as a special case, it would return null from the local cache.

In the end, I chose to solve them with my own use case which is not generalized.
